File: ruby/triangle/triangle.rb
class Triangle
  
  TRIANGLE_KINDS = %i[equilateral isosceles scalene]

  def initialize(a, b, c)
    TriangleValidator.new(a, b, c).validate!
    @a = a
    @b = b
    @c = c
  end

  def kind
    TRIANGLE_KINDS.detect { |triangle_kind| send("#{triangle_kind}?") }
  end

  def equilateral?
    sides.uniq.count == 1
  end

  def isosceles?
    sides.uniq.count == 2
  end

  def scalene?
    sides.uniq.count == 3
  end

  private

    def sides
      [@a, @b, @c]
    end

end

class TriangleValidator

  def initialize(a, b, c)
    @a = a
    @b = b
    @c = c
  end

  def validate!
    unless triangle_inequality_satisfied?
      raise TriangleError, "Invalid sides' values - cannot create a triangle"
    end
  end

  def triangle_inequality_satisfied?
    (@b - @c).abs < @a and @a < (@b + @c)
  end

end

class TriangleError < Exception
end

File: python/triangle/triangle.py
class Triangle(object):
  
    def __init__(self, a, b ,c):
        self.a = a 
        self.b = b 
        self.c = c 
        if not self._validate():
            raise TriangleError

    def kind(self):
        for kind, result in self._kinds.items():
            if kind():
                return result
        else: 
            return "scalene"

    def is_equilateral(self):
        return len(set(self._sides)) == 1

    def is_isosceles(self):
        return len(set(self._sides)) == 2

    def _validate(self):
        return abs(self.a - self.c) < self.b and self.b < self.a + self.c

    @property
    def _kinds(self):
        return {
            self.is_equilateral: "equilateral",
            self.is_isosceles: "isosceles"
        }

    @property
    def _sides(self):
        return [self.a, self.b, self.c]
    

class TriangleError(Exception):
  
    def __init__(self):
        pass